src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -2696,7 +2743,118 @@ void HwEmShim::closemMessengerThread() {
 		mMessengerThread.join();
 		mMessengerThreadStarted = false;
 	}
+    if(mHostMemAccessThreadStarted) {
+      mHostMemAccessThread.join(); 
+      mHostMemAccessThreadStarted = false;
+    }
 }
 /********************************************** QDMA APIs IMPLEMENTATION END**********************************************/
 /**********************************************HAL2 API's END HERE **********************************************/
+
+/********************************************** Q2H_helper class implementation starts **********************************************/
+bool device2xrt_trans_cb(bool, int, unsigned long int, void*,unsigned long int) {
+    return true;
+}
+Q2H_helper :: Q2H_helper() : buf_size(0),buf(NULL), r(0) {
+    ci_msg = new call_packet_info;
+    ri_msg = new response_packet_info;
+    Q2h_sock = new unix_socket("xcl_sock_K2H");
+    ci_msg->set_size(0);
+    ci_msg->set_xcl_api(0);
+    ri_msg->set_size(0);
+    ri_msg->set_xcl_api(0);
+    i_len = ci_msg->ByteSize();
+    i_buf = malloc(i_len);
+    ri_len = ri_msg->ByteSize();
+    ri_buf = malloc(i_len);
+}
+Q2H_helper::~Q2H_helper() {
+    if(buf != NULL)
+        free(buf);
+    free(i_buf);
+    free(ri_buf);
+    delete ci_msg;
+    delete ri_msg;
+    delete Q2h_sock;
+}
+int Q2H_helper::poolingon_Qdma() {
+    //Pooling on socket for any memory or interrupt requests from SIM_QDMA
+    memset(i_buf,0,i_len);
+    r = Q2h_sock->sk_read(i_buf, i_len);
+    if (r <= 0) {
+    	return r;
+    }
+    assert(i_len == (uint32_t)r);
+    ci_msg->ParseFromArray(i_buf, i_len);
+    if (ci_msg->xcl_api() == xclClose_n) {
+        return -1;
+    }
+    if (buf_size == 0) {
+    	buf = malloc(ci_msg->size());
+        buf_size = ci_msg->size();
+    } else if (buf_size < ci_msg->size()) {
+    	buf = (void*) realloc(buf, ci_msg->size());
+        buf_size = ci_msg->size();
+    }
+    memset(buf,0,buf_size);
+    r = Q2h_sock->sk_read(buf, ci_msg->size());
+    assert((uint32_t)r == ci_msg->size());
+    select_msg();
+    return 1;
+}
+void Q2H_helper::select_msg() {
+    if (ci_msg->xcl_api() == xclQdma2HostReadMem_n) {
+    	xclSlaveReadReq_call c_msg;
+    	c_msg.ParseFromArray(buf, r);
+        void *data = malloc(c_msg.size());
+        bool resp = (*m_q2h_slavetransaction)(false, 0,(unsigned long int)c_msg.addr(),data,(unsigned long int)c_msg.size());
+        xclSlaveReadReq_response r_msg;
+        r_msg.set_valid(resp);
+        r_msg.set_data(data,c_msg.size());
+        int r_len = r_msg.ByteSize();
+        SEND_RESP2QDMA();
+        free(data);
```

Comment:
This is not good code.   Never ever use
```
setup();
doSomething();
cleanup();
```
Any object that needs to be cleaned up (freed) should be managed by another object that does the cleanup.  The code is not exception safe.

Reply:
Tried to cleanup this code as well.